File: apps/mobile/modules/pulse-fl-executorch/ios/PulseFlExecuTorchModule.swift
import ExpoModulesCore

public class PulseFlExecuTorchModule: Module {
  // Swift module definition
  public func definition() -> ModuleDefinition {
    Name("PulseFlExecuTorch")

    // Loads the model compiled binary (.pte) into memory
    Function("loadModel") { (modelPath: String) -> Bool in
      // IN PRODUCTION:
      // swift can bridge directly to C++ using:
      // #include <executorch/extension/module/module.h>
      // auto model = torch::executor::Module(modelPath);
      // Store in memory cache
      NSLog("PulseFlExecuTorch [iOS]: Loading model from \(modelPath)")
      return true
    }

    // Runs a forward pass on ECG input values
    AsyncFunction("runInference") { (inputData: [Float], promise: Promise) in
      // IN PRODUCTION:
      // auto result = model.forward(tensor_inputs);
      // resolve with logits array
      NSLog("PulseFlExecuTorch [iOS]: Running forward pass on \(inputData.count) parameters")
      promise.resolve([0.05, 0.95] as [Float])
    }

    // Runs local backpropagation epochs on patient data and writes updated safetensors
    AsyncFunction("trainModel") { (globalSafetensorsUri: String, inputs: [[Float]], labels: [Int], outputUri: String, promise: Promise) in
      // IN PRODUCTION:
      // Run local training steps using ExecuTorch gradients or standard C++ optimizer,
      // serialize state dict and write to outputUri.
      NSLog("PulseFlExecuTorch [iOS]: Fine-tuning model on \(inputs.count) local samples...")
      
      // Simulate small training duration
      DispatchQueue.global(qos: .userInitiated).asyncAfter(deadline: .now() + 1.5) {
        // Return dummy final training loss
        promise.resolve(0.038)
      }
    }
  }
}

File: apps/mobile/modules/pulse-fl-executorch/android/src/main/java/com/pulsefl/executorch/PulseFlExecuTorchModule.kt
package com.pulsefl.executorch

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class PulseFlExecuTorchModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("PulseFlExecuTorch")

    // Loads the model file on Android
    Function("loadModel") { modelPath: String ->
      // JNI implementation:
      // System.loadLibrary("executorch");
      // JniLoader.loadModel(modelPath);
      android.util.Log.d("PulseFlExecuTorch", "Loading ExecuTorch model from $modelPath")
      true
    }

    // Runs forward inference returning prediction logits
    AsyncFunction("runInference") { inputData: DoubleArray ->
      android.util.Log.d("PulseFlExecuTorch", "Running inference on ${inputData.size} data points")
      doubleArrayOf(0.08, 0.92)
    }

    // Runs local backpropagation training iterations
    AsyncFunction("trainModel") { globalSafetensorsUri: String, inputs: Array<DoubleArray>, labels: IntArray, outputUri: String ->
      android.util.Log.d("PulseFlExecuTorch", "Starting on-device backpropagation training round")
      
      // Simulate local training step duration (1.5 seconds)
      delay(1500)
      
      // Return final local loss
      0.042
    }
  }
}
